data-processing/entities/citations/commands/upload_citations.py

```diff
@@ -104,30 +104,31 @@ def _entity_data_from_text_and_matches(
         self,
         arxiv_id: ArxivId,
         bibitem_texts: Dict[str, str],
-        key_s2_ids: Dict[str, str],
+        key_s2_ids: Dict[str, S2Id],
+        s2_data: Dict[S2Id, SerializableReference],
         bibitem_key: str,
     ) -> EntityData:
         entity_data: EntityData = {
             "key": bibitem_key,
         }
 
+        def log_missing(missing: str):
+            logging.warning(f"Missing {missing} for bibitem with key {bibitem_key} for paper {arxiv_id}")
+
         if bibitem_key in bibitem_texts:
             entity_data["bibitem_text"] = bibitem_texts[bibitem_key]
         else:
-            logging.warning(
-                "Missing bibitem text for bibitem with key %s for paper %s",
-                bibitem_key,
-                arxiv_id,
-            )
+            log_missing("bibitem text")
 
         if bibitem_key in key_s2_ids:
-            entity_data["paper_id"] = key_s2_ids[bibitem_key]
+            reference_s2_id = key_s2_ids[bibitem_key]
+            entity_data["paper_id"] = reference_s2_id
+            if reference_s2_id in s2_data:
+                entity_data["corpus_id"] = s2_data[reference_s2_id].corpus_id
```

Comment:
So s2_data is reference info keyed by the legacy paper id of the reference? So s2_data only contains info about matched references?

Reply:
* `ResolveBibItems` make matches by comparing title strings of the PDF bib entry VS the reference titles given by Public API.  
  * INTERESTING!!!  The old API **ONLY** provides references that has a paper_id/SHA.  
    * but **NOT** the new API!
    * So given that not all the reference items from Public API actually have paper_id/SHA or any other ids,
      * ...it's entirely possible for the `ResolveBibItems` step to make a `BibitemMatch` for a bib entry to such a paper!
      * But we shouldn't even allow such a match to occur in the first place bc there's no point!  
      * OK, I'm gonna need to change soemthing in `ResolveBibItems` to exclude from matching on these deadbeat references from PublicAPI!!!
  * These `BibitemMatch` results are saved to the intermediate `resolution.csv`
* So this next step `UploadCitations` grabs that `resolutions.csv`
  * L125 could actually be None, oh no!  (Hence the fix to exclude garbage above!)
  
STAY TUNED!!! still thinking/tinkering...